x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/RestAuthenticateAction.java

```diff
@@ -65,7 +70,16 @@ public RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient c
                 new RestBuilderListener<AuthenticateResponse>(channel) {
             @Override
             public RestResponse buildResponse(AuthenticateResponse authenticateResponse, XContentBuilder builder) throws Exception {
-                authenticateResponse.authentication().toXContent(builder, ToXContent.EMPTY_PARAMS);
+                if (anonymousEnabled
+                    && false == anonymousUser.equals(authenticateResponse.authentication().getUser())
+                    && anonymousUser.roles().length != 0) {
+                    builder.startObject();
+                    authenticateResponse.authentication().toXContentFragment(builder);
+                    builder.array(User.Fields.ANONYMOUS_ROLES.getPreferredName(), anonymousUser.roles());
+                    builder.endObject();
+                } else {
+                    authenticateResponse.authentication().toXContent(builder, ToXContent.EMPTY_PARAMS);
+                }
```

Comment:
The tricky bits of this work are:
* XContent serialization is managed inside Authentication
* Information about anonymous user is _not_ available in Authentication.

One approach is to add an new `anonymousRoles` field to `User` class and populate it during authentication. But this has wide spread impact as the `User` object is serialised and sent everywhere. Also the anonymous access is by default off, so it does not seem to justify the wide spread change.

So the new logic is added here in `RestAuthenticationAction`, which can create the anonymous user object and inject its roles into the XContent object.

Reply:
> One approach is to add an new anonymousRoles field to User class and populate it during authentication. 

I think the best way forward is to actually do this ( no need to add a field to `User`), but evaluate the anonymous roles during authentication (i.e. right after the realms return the User in `AuthenticationService#consumeToken` ) instead of during authorization. 

If we don't want to do this now, but defer it to a later point ( or as part of the effort to refactor the logic around `AuthenticationToken`s ), I think this fits better in the Transport Action than in the Rest one.

Reply:
I think this will only work if we move away from adding the new `anonymous_role` field in the response, am I right? 
Otherwise, there is no place to save these role information. We could use `User#metadata` but it feels hacky.

Reply:
Yes this implies we follow my other comment and anonymous roles are part of roles

Reply:
Great. Thanks for the clarification.

Reply:
@jkakavas There are some issues for adding the anonymous roles into User#roles
* `AuthenticationService#consumeToken` does not cover runAs user, which is handled in `lookupRunAsUser`.
* Both above methods end in `finishAuthentication` , so this could be a better candidate.
* But doing above changes the existing logic. It's adding the anonymous roles to all users. Currently, some system users do not inherit anonymous roles. The code even [asserts that XPackUser can only have a single role](https://github.com/elastic/elasticsearch/blob/7.6/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L207). There could be other situations where adding anonymous roles are not desired?
* The final role resolving is handled in `CompositeRolesStore#getRoles`. Result of this method is the source of truth as it is what gets used in execution.

In summary, adding roles to all authenticated user has wider impact and it does not feel ideal. I think that the Authorization process may still be a better place for this, because "granting more roles" to user sounds more like "authorization" instead of "authentication". In fact, `CompositeRolesStore#getRoles` (part of authorization) is where this happens.

The challenge of using result of `CompositeRolesStore#getRoles` is that it is not easily accessible:

* The resulted AuthorizationInfo object does get saved in threadContext as a *Transient* value. But this does not seem to be accessible once cross the wire.
* It is also not possible (or ideal) to alter the `User` object in authorization process for saving these resolved roles.
* Adding a new key in threadContext for saving these roles seem to be overkill and inconsistent.

So I'd like to propose the following approach:
* Add a new field for the resolved roles in  `org.elasticsearch.xpack.core.security.action.user.AuthenticateResponse`. 
* In `TransportAuthenticateAction`, grab the resolved roles from threadContext and set it to the `AuthenticateResponse` object. It may be possible to access the transient value here because the transport action runs in the same machine as the "authorization" process, i.e. not cross wire.
* `RestAuthenticateAction` has access to `AuthenticateResponse` from which the final response can be built.

Please let me know your thoughts. Thanks.

Reply:
> There could be other situations where adding anonymous roles are not desired?

I can think of reserved and internal users. I don't think we need to care about internal users here ( `_system`, `_xpack` etc ) as we wouldn't be handling requests from the REST layer as these users. But reserved users ( i.e. `kibana` ) would be problematic. 

>  I think that the Authorization process may still be a better place for this, because "granting more roles" to user sounds more like "authorization" instead of "authentication". In fact, CompositeRolesStore#getRoles (part of authorization) is where this happens.

But we do resolve and populate the user roles in authentication too which makes it hard to have this mental distinction. I would argue that role resolving based on user attributes ( i.e. role mapping ) is part of authentication. Granting access based on the definition of these roles is authorization. 

I still feel that this fits better in AuthenticationService but did not have the cycles to think it through. I think this would benefit from a short live discussion, care to put it on the agenda for our weekly meeting tomorrow? We can reach a consensus there and come back here with that and see this through to completion

Reply:
I added it under Forward agenda. Thanks.